sefaria/model/portal.py

```diff
@@ -0,0 +1,121 @@
+
+
+class InvalidURLException(Exception):
+    def __init__(self, url):
+        self.url = url
+        self.message = f"'{url}' is not a valid URL."
+        super().__init__(self.message)
+
+def validate_url(url):
+    try:
+        # Attempt to parse the URL
+        validator = URLValidator()
+        validator(url)
+        return True
+
+    except ValidationError:
+        # URL parsing failed
+        raise InvalidURLException(url)
+class InvalidHTTPMethodException(Exception):
+    def __init__(self, method):
+        self.method = method
+        self.message = f"'{method}' is not a valid HTTP API method."
+        super().__init__(self.message)
+
+def validate_http_method(method):
+    """
+    Validate if a string represents a valid HTTP API method.
+
+    Args:
+        method (str): The HTTP method to validate.
+
+    Raises:
+        InvalidHTTPMethodException: If the method is not valid.
+
+    Returns:
+        bool: True if the method is valid, False otherwise.
+    """
+    valid_methods = ["GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS", "HEAD"]
+
+    # Convert the method to uppercase and check if it's in the list of valid methods
+    if method.upper() in valid_methods:
+        return True
+    else:
+        raise InvalidHTTPMethodException(method)
+
+class Portal(abst.SluggedAbstractMongoRecord):
+    collection = 'portals'
+    slug_fields = ['slug']
+
+    required_attrs = [
+        "slug",
+        "about",
+    ]
+    optional_attrs = [
+        "mobile",
+        'newsletter',
+    ]
+
+    def _validate(self):
+        super(Portal, self)._validate()
+
+        about_schema = {
+            "title": ({"en": (str, "required"), "he": (str, "required")}, "required"),
+            "title_url": (str, "optional"),
+            "image_uri": (str, "optional"),
+            "image_caption": ({"en": (str, "optional"), "he": (str, "optional")}, "optional"),
+            "description": ({"en": (str, "required"), "he": (str, "required")}, "optional"),
+        }
+
+        mobile_schema = {
+            "title": ({"en": (str, "required"), "he": (str, "required")}, "required"),
+            "description": ({"en": (str, "optional"), "he": (str, "optional")}, "optional"),
+            "android_link": (str, "optional"),
+            "ios_link": (str, "optional")
+        }
+
+        newsletter_schema = {
+            "title": ({"en": (str, "required"), "he": (str, "required")}, "required"),
+            "title_url": (str, "optional"),
+            "description": ({"en": (str, "required"), "he": (str, "required")}, "optional"),
+            "api_schema": ({"http_method": (str, "required"),
+                            "payload": ({"first_name_key": (str, "optional"), "last_name_key": (str, "optional"), "email_key": (str, "optional")}, "optional")}
+                           , "optional")
+        }
+
+        if hasattr(self, "about"):
+            abst.validate_dictionary(self.about, about_schema)
```

Comment:
Why make them a module level method if you just need to use these validators as inherited methods?

Reply:
This method feels generic enough that it can apply to other classes besides AbstractMongoRecord.

Reply:
But then it doesnt get hooked into the normal flow of the validate methods on the object?